### PR TITLE
azblob: Make azblob support private cloud endpoint suffix

### LINKED ISSUE #6863: Fail to connect azure stack hub's azure blob storage
**Describe the bug**
azblob do not consider endpoint configuration when select blob/container URL if the storage is not on azure public cloud (ex: azure stack hub)

loki version: main-19c7315
promtail version: 2.6.1

The endpoint suffix of azure blob storage of Azure Stack Hub is self defined by company. 
From the code snap, we could see that we could configure azure storage endpoint suffix for azblob and getServicePrincipleToken() will refer this configuration.  

    func (b *BlobStorage) getServicePrincipalToken() (*adal.ServicePrincipalToken, error) {
        var endpoint string
        if b.cfg.Endpoint != "" {
            endpoint = b.cfg.Endpoint
        } else {
            endpoint = defaultEndpoints[b.cfg.Environment]
        }

But when select blob/container URL, they does not refer this configuration and cause connection to azure stack hub blob storage fail.

    func (b *BlobStorage) selectBlobURLFmt() string {
        return fmt.Sprintf("https://%%s.%s/%%s/%%s", defaultEndpoints[b.cfg.Environment])
    }

**To Reproduce**
1. Started Loki main-19c7315
2. Started Promtail (2.6.1) to tail '...'
3. Use azure blob storage of azure stack hub as loki's storage

**Expected behavior**
Connection should be successful and logs could be save to azure blob storage of azure stack hub

**Environment:**
 - Infrastructure: k8s
 - Deployment tool: helm

**Screenshots, Promtail config, or terminal output**
If applicable, add any output to help explain your problem.
loki-config

### PR DESCRIPTION
**Pull request by sardismart**
**What this PR does / why we need it**:  
The endpoint suffix of private cloud(azure stack hub) azblob is different from Azure. The loki configuration has already support customize endpoint suffix but not use when select blob/container URL.  
This PR introduce following change:  
* Consider endpoint suffix from configuration when select blob or container URL format

**Which issue(s) this PR fixes**:
Fixes #6863

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [x] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

**Comment by sardismart** [comment]
This is to solve
https://github.com/grafana/loki/issues/6863

**Review by dannykopping — COMMENTED** [review]
This looks reasonable to me @sardismart, but can you please add a test and some documentation?

**Comment by sardismart** [comment]
> This looks reasonable to me @sardismart, but can you please add a test and some documentation?

Is there any examples that I can follow ?

File: pkg/storage/chunk/client/azure/blob_storage_client_test.go [hunk 1]
```diff
@@ -92,3 +93,29 @@ func Test_Hedging(t *testing.T) {
 		})
 	}
 }
+
+func Test_EndpointSuffixWithContainer(t *testing.T) {
```
Comment (dannykopping) [inline pkg/storage/chunk/client/azure/blob_storage_client_test.go:97]: This looks great, but it only tests the new changes - can you also add a test to validate that the default address still works please?

Comment (sardismart) [reply pkg/storage/chunk/client/azure/blob_storage_client_test.go:108]: updated

Comment (dannykopping) [reply pkg/storage/chunk/client/azure/blob_storage_client_test.go:108]: Thank you

**Review by dannykopping — APPROVED** [review]
LGTM, thanks for your contribution @sardismart!
I'll try fix the CLA now, then merge

**Comment by dannykopping** [comment]
@sardismart please try sign the CLA again, or show me what you're seeing when you try to and it says it's signed?
I tried to retrigger the CI check but it's failing to change state.

**Comment by sardismart** [comment]
@dannykopping 
I use CLA assistant and it shows I have agreed
Is it possible that caused by my first commit which without link my email address?

image

**Comment by dannykopping** [comment]
> @dannykopping I use CLA assistant and it shows I have agreed Is it possible that caused by my first commit which without link my email address?
> 
> image

OK, thanks - I'll merge :+1:

**Comment by sardismart** [comment]
@dannykopping Just curious when will you merge this PR?

**Comment by dannykopping** [comment]
> @dannykopping Just curious when will you merge this PR?

I don't have permission to override the CLA check. I've asked @slim-bean to merge it

**Comment by sardismart** [comment]
@dannykopping  I have rebase my branch and seems the CLA check is passed

**Pull request merged**